### PR TITLE
Update to fo-dicom 5

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
Update to `fo-dicom` 5.x and remove `Json.NET` dependency

- [x] Update fo-dicom to `5.0.2`
  - [x] Update namespaces
  - [x] Integrate new DI APIs
  - [x] Replace calls to `DicomValidation` with `DicomValidationBuilderExtension`
  - [x] Address API changes, like the assumption of UTF-8 encodings
  - [x] Create decorator over the Fellow Oak logger that logs to our own log sink via DI
- [x] Migrate to fo-dicom.codecs over efferent
  - [x] Target `x64` for all assemblies due to dependency on native `x64` codec
- [x] Migrate DICOM service to `System.Text.Json` over `Json.NET`
  - [x] Remove dependency on `fo-dicom.json` (new serializers are present in the `fo-dicom` packages)
  - [x] Replace `JsonConvert` with `JsonSerializer`
  - [x] Replace converters
  - [x] Pass serializer options via DI to services
  - [x] Ensure consistent object contracts for deserialization (eg. Via property setters or ctors)
  - [x] Keep `Json.NET` for Azure Functions as it is used internally

## Related issues
[AB#85800](https://microsofthealth.visualstudio.com/Health/_workitems/edit/85800)

## Testing
Locally and via pipeline

### CODE REVIEW

**Comment by krushnavadan** [comment]
i am working on same if you wants to connect

**Comment by wsugarman** [comment]
Unfortunately, we cannot check this change in yet as there is no version of the fo-dicom 5+ library without issues.
- `5.0.0` cannot serialize and deserialize datasets properly
    - https://github.com/fo-dicom/fo-dicom/issues/1251
- `5.0.1` has issues with the new loading/saving logic
    - https://github.com/fo-dicom/fo-dicom/issues/1264

Once a new version of fo-dicom is published, we can integrate it safely.

**Comment by krushnavadan** [comment]
Thanks for the reply, but on the current version we are facing an issue with partially image load, any suggestion and help is appreciated

File: src/Microsoft.Health.Dicom.Core/Features/Retrieve/Transcoder.cs [hunk 1]
```diff
@@ -34,13 +33,8 @@ public class Transcoder : ITranscoder
 
         public Transcoder(RecyclableMemoryStreamManager recyclableMemoryStreamManager, ILogger<Transcoder> logger)
         {
-            EnsureArg.IsNotNull(recyclableMemoryStreamManager, nameof(recyclableMemoryStreamManager));
-            EnsureArg.IsNotNull(logger, nameof(logger));
-            _recyclableMemoryStreamManager = recyclableMemoryStreamManager;
-            _logger = logger;
-
-            // Use Efferent transcoder
-            TranscoderManager.SetImplementation(new NativeTranscoderManager());
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Retrieve/Transcoder.cs:43]: Moved to DI

File: src/Microsoft.Health.Dicom.Core/Features/Store/Entries/StreamOriginatedDicomInstanceEntry.cs [hunk 2]
```diff
@@ -34,36 +33,34 @@ internal StreamOriginatedDicomInstanceEntry(Stream seekableStream)
...
         /// <inheritdoc />
         public async ValueTask<DicomDataset> GetDicomDatasetAsync(CancellationToken cancellationToken)
         {
-            if (Interlocked.CompareExchange(ref _dicomFileLoadedState, 1, 0) == 0)
+            try
             {
-                // Load the file.
-                _dicomFile = await DicomFile.OpenAsync(_stream, FileReadOption.SkipLargeTags);
-                _dicomFileLoadingCompletionSource.SetResult(true);
+                DicomFile file = await _dicomFileCache.GetAsync(cancellationToken: cancellationToken);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Features/Store/Entries/StreamOriginatedDicomInstanceEntry.cs:44]: Some re-use while I was here

File: src/Microsoft.Health.Dicom.Core/Logging/FellowOakLoggerDecorator.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+using System.Diagnostics.CodeAnalysis;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+using FellowOakLogLevel = FellowOakDicom.Log.LogLevel;
+using IFellowOakLogger = FellowOakDicom.Log.ILogger;
+
+namespace Microsoft.Health.Dicom.Core.Logging
+{
+    [SuppressMessage("Usage", "CA2254:Template should be a static expression", Justification = "Class serves as a passthrough for ILogger arguments.")]
+    internal class FellowOakLoggerDecorator : IFellowOakLogger
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Core/Logging/FellowOakLoggerDecorator.cs:16]: For some reason `fo-dicom` has its own logger

File: src/Microsoft.Health.Dicom.Api/Registration/DicomServerApplicationBuilderExtensions.cs [hunk 4]
```diff
@@ -33,8 +34,10 @@ public static IApplicationBuilder UseDicomServer(this IApplicationBuilder app)
 
             app.UseHealthChecksExtension(new PathString(KnownRoutes.HealthCheck));
 
-            IOptions<FeatureConfiguration> featureConfiguration = app.ApplicationServices.GetRequiredService<IOptions<FeatureConfiguration>>();
+            // Update Fellow Oak DICOM services to use ASP.NET Core's service container
+            DicomSetupBuilder.UseServiceProvider(app.ApplicationServices);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/Registration/DicomServerApplicationBuilderExtensions.cs:38]: I am unsure if there is a better place to hook this in. We need to configure `fo-dicom` to use our service container.

File: src/Microsoft.Health.Dicom.Blob/Features/Storage/BlobMetadataStore.cs [hunk 5]
```diff
@@ -142,11 +129,11 @@ private BlockBlobClient GetInstanceBlockBlob(VersionedInstanceIdentifier version
             return _container.GetBlockBlobClient(blobName);
         }
 
-        private static async Task ExecuteAsync(Func<Task> action)
+        private static async Task<T> ExecuteAsync<T>(Func<CancellationToken, Task<T>> action, CancellationToken cancellationToken)
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Blob/Features/Storage/BlobMetadataStore.cs:132]: A bit of refactoring while I was here

File: src/Microsoft.Health.Dicom.Api/Features/Formatters/DicomJsonOutputFormatter.cs [hunk 6]
```diff
@@ -1,90 +0,0 @@
...
-namespace Microsoft.Health.Dicom.Api.Features.Formatters
-{
-    public class DicomJsonOutputFormatter : TextOutputFormatter
-    {
-        private readonly JsonDicomConverter _jsonDicomConverter = new JsonDicomConverter(writeTagsAsKeywords: false);
-        private readonly JsonSerializer _jsonSerializer = new JsonSerializer();
-
-        public DicomJsonOutputFormatter()
-        {
-            SupportedMediaTypes.Add(KnownContentTypes.ApplicationDicomJson);
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Api/Features/Formatters/DicomJsonOutputFormatter.cs:30]: This formatter triggers for a subset of the JSON media type, and therefore the default json output formatter should trigger and use the configured converter from `fo-dicom`

File: Directory.Build.props [hunk 7]
```diff
@@ -10,13 +10,15 @@
...
     <FhirServerPackageVersion>2.0.55</FhirServerPackageVersion>
-    <FoDicomVersion>4.0.8</FoDicomVersion>
+    <FoDicomVersion>5.0.2</FoDicomVersion>
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <HealthcareSharedPackageVersion>4.0.8</HealthcareSharedPackageVersion>
     <HighEntropyVA>true</HighEntropyVA>
     <Hl7FhirPackageVersion>3.6.0</Hl7FhirPackageVersion>
     <LangVersion>Latest</LangVersion>
     <PackageLicenseExpression>MIT</PackageLicenseExpression>
+    <Platforms>x64</Platforms>
```
Comment (wsugarman) [inline Directory.Build.props:20]: TODO: Add reason why

**Review by smithago — APPROVED** [review]

**Pull request merged**